volume/guest_image.go

```diff
@@ -0,0 +1,223 @@
+type GuestVolumeImageOutput struct {
+	Error string
+}
+
+// FromGuestImage returns a new provider to that expose the volumes on the given image.
+func FromGuestImage(client *containerd.Client, vmID, image, snapshot string, volumes []string, opts ...ImageOpt) *GuestVolumeImageProvider {
+	p := GuestVolumeImageProvider{
+		client:   client,
+		image:    image,
+		snapshot: snapshot,
+		vmID:     vmID,
+		volumes:  volumes,
+	}
+	p.config.snapshotter = containerd.DefaultSnapshotter
+
+	for _, opt := range opts {
+		opt(&p.config)
+	}
+
+	return &p
+}
+
+// Name of the provider.
+func (p *GuestVolumeImageProvider) Name() string {
+	return p.image
+}
+
+// pull the image.
+func (p *GuestVolumeImageProvider) pull(ctx context.Context) error {
+	remoteOpts := []containerd.RemoteOpt{
+		containerd.WithPullUnpack,
+		containerd.WithPullSnapshotter(p.config.snapshotter),
+	}
+	remoteOpts = append(remoteOpts, p.config.pullOpts...)
+	image, err := p.client.Pull(ctx, p.image, remoteOpts...)
+	if err != nil {
+		return err
+	}
+
+	p.cImage = image
+	return nil
+}
+
+func mountHostDirs(image string) oci.SpecOpts {
+	return func(ctx context.Context, client oci.Client, container *containers.Container, s *oci.Spec) error {
+		s.Mounts = append(s.Mounts,
+			specs.Mount{
+				Source:      vmBinDir,
+				Destination: containerBinDir,
+				Type:        "bind",
+				Options:     []string{"bind", "ro"},
+			},
+			specs.Mount{
+				Source:      vmVolumesDir,
+				Destination: containerVolumesDir,
+				Type:        "bind",
+				Options:     []string{"bind"},
+			},
+		)
+		return nil
+	}
+}
+
+// copy files from the image by launching the container.
+// The name must be unique within the VM and must be /[A-Z0-9a-z][A-Z0-9a-z._-]*/.
+func (p *GuestVolumeImageProvider) copy(ctx context.Context, containerName string) error {
+	err := identifiers.Validate(containerName)
+	if err != nil {
+		return err
+	}
+	container, err := p.client.NewContainer(ctx,
+		containerName,
+		containerd.WithSnapshotter(p.config.snapshotter),
+		containerd.WithNewSnapshot(p.snapshot, p.cImage),
+		containerd.WithNewSpec(
+			firecrackeroci.WithVMID(p.vmID),
+			oci.WithProcessArgs(filepath.Join(containerBinDir, "volume-init")),
+			oci.WithProcessCwd("/"),
+			mountHostDirs(containerName),
+		),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create container %q with %q: %w", containerName, p.config.snapshotter, err)
+	}
+	defer container.Delete(ctx, containerd.WithSnapshotCleanup)
+
+	input := GuestVolumeImageInput{From: "/", To: filepath.Join(containerVolumesDir, containerName), Volumes: p.volumes}
+	b, err := json.Marshal(&input)
+	if err != nil {
+		return err
+	}
+
+	stdin := bytes.NewBuffer(b)
+	var stdout, stderr bytes.Buffer
+	task, err := container.NewTask(ctx, cio.NewCreator(cio.WithStreams(stdin, &stdout, os.Stderr)))
+	if err != nil {
+		return err
+	}
+	defer task.Delete(ctx)
+
+	exitCh, err := task.Wait(ctx)
+	if err != nil {
+		return err
+	}
```

Comment:
this is correct, should begin to wait for task before starting it to avoid races. See https://github.com/containerd/containerd/blob/main/docs/getting-started.md#task-wait-and-start

> This makes sure that you do not encounter any races if the task has a simple program like /bin/true that exits promptly after calling start.